mmf/models/mmbt.py

```diff
@@ -352,6 +352,43 @@ def forward(self, sample_list):
         if self.config.use_modal_end_token:
             modal_end_token = sample_list.input_ids[:, -1].clone().detach()
 
+        if hasattr(sample_list, "modal_token_type_ids"):
+            modal_token_type_ids = sample_list.modal_token_type_ids
+        else:
+            modal_token_type_ids = torch.zeros(
+                (input_modal.size(0)),
+                dtype=torch.long,
+                device=input_modal.device,
+            )
+
+            segment_ids = sample_list.segment_ids
+            max_id = segment_ids.max()
+            min_id = segment_ids.min()
+            # Case of only one segment
+            if max_id == min_id:
+                # If max_id is greater than 0, that means text is at 0 segment
+                # which means modal will be at 1
+                # In other case, it will be zero, which it already is
+                if max_id == 0:
+                    modal_token_type_ids = torch.ones(
+                        (input_modal.size(0)),
+                        dtype=torch.long,
+                        device=input_modal.device,
+                    )
+            else:
+                max_segment = self.config.text_encoder.params.get("num_segments", 2) - 1
+                # If max id is not equal to max_segment, it means
+                # text segments start from 0 which means modal will
+                # be last, otherwise, it is 0, which is already is
+                if max_id != max_segment:
+                    modal_token_type_ids = torch.full(
+                        (input_modal.size(0), 1),
+                        fill_value=max_segment,
+                        dtype=torch.long,
+                        device=input_modal.device,
+                    )
+
```

Comment:
This can probably be refactored to first determine the proper value of the modal segment, and then set it using `torch.full` once (instead of using torch.zeros, torch.ones for different use cases).

```
modal_token_type_ids = torch.full(
                        (input_modal.size(0), 1),
                        fill_value=max_segment,
                        dtype=torch.long,
                        device=input_modal.device,
                    )
```

Reply:
Refactored. Thanks for the suggestion.